auth/src/main/java/feast/auth/credentials/GoogleAuthCredentials.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright 2018-2020 The Feast Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.auth.credentials;
+
+import static io.grpc.Metadata.ASCII_STRING_MARSHALLER;
+
+import com.google.auth.oauth2.IdTokenCredentials;
+import com.google.auth.oauth2.ServiceAccountCredentials;
+import io.grpc.CallCredentials;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.Executor;
+
+/*
+ * Google auth provider's callCredentials Implementation for serving.
+ * Used by CoreSpecService to connect to core.
+ */
+public class GoogleAuthCredentials extends CallCredentials {
+  private String accessToken;
+  private Instant tokenExpiryTime;
+  private final IdTokenCredentials credentials;
+  private static final String BEARER_TYPE = "Bearer";
+  private static final Metadata.Key<String> AUTHORIZATION_METADATA_KEY =
+      Metadata.Key.of("Authorization", ASCII_STRING_MARSHALLER);
+
+  public GoogleAuthCredentials(Map<String, String> options) throws IOException {
+
+    String targetAudience = options.getOrDefault("audience", "https://localhost");
+    ServiceAccountCredentials serviceCreds =
+        (ServiceAccountCredentials)
+            ServiceAccountCredentials.getApplicationDefault()
+                .createScoped(Arrays.asList("openid", "email"));
+
+    credentials =
+        IdTokenCredentials.newBuilder()
+            .setIdTokenProvider(serviceCreds)
+            .setTargetAudience(targetAudience)
+            .build();
+  }
+
+  @Override
+  public void applyRequestMetadata(
+      RequestInfo requestInfo, Executor appExecutor, MetadataApplier applier) {
+    appExecutor.execute(
+        () -> {
+          try {
+            // Fetches new token if it is not available or if token has expired.
+            if (this.accessToken == null || Instant.now().isAfter(this.tokenExpiryTime)) {
```

Comment:
It's strange that the Google gRPC library doesn't have refreshing the credentials built into it.

Reply:
It has `refreshIfExpired` which I'm using. I will check if that condition is good enough.